sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -2856,6 +2885,20 @@ pi_result piQueueRelease(pi_queue Queue) {
         if (ZeResult && ZeResult != ZE_RESULT_ERROR_UNINITIALIZED)
           return mapError(ZeResult);
       }
+      if (Queue->UsingImmCmdLists && Queue->OwnZeCommandQueue) {
+        std::scoped_lock<ur_mutex> Lock(
+            Queue->Context->ZeCommandListCacheMutex);
+        // Add commandlists to the cache for future use.
+        // They will be deleted when the context is destroyed.
```

Comment:
Isn't this a good place to just destroy the command-list if it cannot be reused (and then we don't need the `pi_command_list_desc_t` structure?

Reply:
A non-reusable IMM can only arise from make_queue, which is very rare. Deleting the IMM here in that case would duplicate the destroy call here, in addition to at context deletion. And you still need the pi_command_list_desc_t struct because inside it is the info about whether the command list is reusable.

Reply:
That's where we normally destroy things, at release (unless want to re-use).
You'd just define the caches as:

>  std::unordered_map<
      ze_device_handle_t,
      std::list<std::pair<ze_command_list_handle_t, **ze_command_queue_desc_t**>>>

Reply:
and since it is only rarely non-reusable, this will save space for many reusable ones in the map

Reply:
The information about whether the command list is reusable or not has to be known here. Where do we get it from? The queue structure? That would mean adding a field to the queue to say its IMMs are not reusable.

Reply:
It would go into `pi_command_list_info_t` which is there exactly for the purpose of providing additional info about the usage of a command-list

Reply:
Ok, that has been done.